validator/keymanager/v2/direct/direct.go

```diff
@@ -320,34 +321,50 @@ func (dr *Keymanager) initializeSecretKeysCache(ctx context.Context) error {
 			}
 		}
 	}()
-	dr.lock.Lock()
-	defer dr.lock.Unlock()
-	_, err = mputil.Scatter(len(accountNames), func(offset int, entries int, _ *sync.RWMutex) (interface{}, error) {
-		for i := 0; i < len(accountNames[offset:offset+entries]); i++ {
-			name := accountNames[i]
-			password, err := dr.wallet.ReadPasswordFromDisk(ctx, name+PasswordFileSuffix)
-			if err != nil {
-				return nil, errors.Wrapf(err, "could not read password for account %s", name)
-			}
-			encoded, err := dr.wallet.ReadFileAtPath(ctx, name, KeystoreFileName)
+
+	keystoreFiles := make([]*v2keymanager.Keystore, len(accountNames))
+	passwords := make([]string, len(accountNames))
+	for i := 0; i < len(accountNames); i++ {
+		name := accountNames[i]
+		encoded, err := dr.wallet.ReadFileAtPath(ctx, name, KeystoreFileName)
+		if err != nil {
+			return errors.Wrapf(err, "could not read keystore file for account %s", name)
+		}
+		keystoreFile := &v2keymanager.Keystore{}
+		if err := json.Unmarshal(encoded, keystoreFile); err != nil {
+			return errors.Wrapf(err, "could not decode keystore file for account %s", name)
+		}
+		password, err := dr.wallet.ReadPasswordFromDisk(ctx, name+PasswordFileSuffix)
+		// If the password is suddenly not found for an account, prompt the user for it.
+		if err != nil && strings.Contains(err.Error(), errFileNotFound) {
+			password, err = dr.promptForAccountPassword(ctx, name, keystoreFile.Pubkey)
 			if err != nil {
-				return nil, errors.Wrapf(err, "could not read keystore file for account %s", name)
-			}
-			keystoreFile := &v2keymanager.Keystore{}
-			if err := json.Unmarshal(encoded, keystoreFile); err != nil {
-				return nil, errors.Wrapf(err, "could not decode keystore file for account %s", name)
+				return err
 			}
+		} else if err != nil {
+			return errors.Wrapf(err, "could not read password for account %s", name)
+		}
+		keystoreFiles[i] = keystoreFile
+		passwords[i] = password
+	}
+	decryptor := keystorev4.New()
+	_, err = mputil.Scatter(len(keystoreFiles), func(offset int, entries int, lock *sync.RWMutex) (interface{}, error) {
+		lock.Lock()
+		defer lock.Unlock()
```

Comment:
Doesn't this defeat the purpose of the scatter? Nothing can be done in parallel if a single routing locks all of the others.